Patch-Source: https://github.com/quick-lint/quick-lint-js/commit/a2798b35021f34bc798e2b70ec703075dd5eb7f6
---
From a2798b35021f34bc798e2b70ec703075dd5eb7f6 Mon Sep 17 00:00:00 2001
From: strager <s@strager.net>
Date: Sun, 15 Jun 2025 18:25:58 -0400
Subject: [PATCH] fix(port): fix bad reference to 'environ' global

We forward-declare 'environ' on Linux. However, our forward declaration is
inside a namespace, and that seems to cause problems on Arch Linux now:

```
/usr/sbin/ld: ../src/libquick-lint-js-lib.a(child-process.cpp.o): warning: relocation against `_ZN13quick_lint_js12_GLOBAL__N_17environE' in read-only section `.text._ZN13quick_lint_js12_GLOBAL__N_1L21get_posix_environmentEv'
/usr/sbin/ld: ../src/libquick-lint-js-lib.a(child-process.cpp.o): in function `quick_lint_js::(anonymous namespace)::get_posix_environment()':
/__w/quick-lint-js/quick-lint-js/src/quick-lint-js/port/child-process.cpp:364:(.text._ZN13quick_lint_js12_GLOBAL__N_1L21get_posix_environmentEv+0x7): undefined reference to `quick_lint_js::(anonymous namespace)::environ'
/usr/sbin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
```

Fix this issue by moving the forward declaration of 'environ' out of a
namespace.
---
 src/quick-lint-js/port/child-process.cpp | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/quick-lint-js/port/child-process.cpp b/src/quick-lint-js/port/child-process.cpp
index 9bb6c76191..70da9034bf 100644
--- a/src/quick-lint-js/port/child-process.cpp
+++ b/src/quick-lint-js/port/child-process.cpp
@@ -42,6 +42,10 @@
 
 using namespace std::literals::string_view_literals;
 
+#if QLJS_HAVE_UNISTD_H && !QLJS_HAVE_NS_GET_ENVIRON
+extern "C" char** environ;
+#endif
+
 namespace quick_lint_js {
 namespace {
 #if QLJS_HAVE_UNISTD_H
@@ -353,10 +357,6 @@ Run_Program_Result run_program(Span<const char* const> command,
 
 namespace {
 #if QLJS_HAVE_UNISTD_H
-#if !QLJS_HAVE_NS_GET_ENVIRON
-extern "C" char** environ;
-#endif
-
 char** get_posix_environment() {
 #if QLJS_HAVE_NS_GET_ENVIRON
   return *::_NSGetEnviron();
